promunifi/clients.go

```diff
@@ -0,0 +1,136 @@
+package promunifi
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"golift.io/unifi"
+)
+
+type uclient struct {
+	Anomalies         *prometheus.Desc
+	BytesR            *prometheus.Desc
+	CCQ               *prometheus.Desc
+	Noise             *prometheus.Desc
+	RoamCount         *prometheus.Desc
+	RSSI              *prometheus.Desc
+	RxBytes           *prometheus.Desc
+	RxBytesR          *prometheus.Desc
+	RxPackets         *prometheus.Desc
+	RxRate            *prometheus.Desc
+	Signal            *prometheus.Desc
+	TxBytes           *prometheus.Desc
+	TxBytesR          *prometheus.Desc
+	TxPackets         *prometheus.Desc
+	TxPower           *prometheus.Desc
+	TxRate            *prometheus.Desc
+	Uptime            *prometheus.Desc
+	WifiTxAttempts    *prometheus.Desc
+	WiredRxBytes      *prometheus.Desc
+	WiredRxBytesR     *prometheus.Desc
+	WiredRxPackets    *prometheus.Desc
+	WiredTxBytes      *prometheus.Desc
+	WiredTxBytesR     *prometheus.Desc
+	WiredTxPackets    *prometheus.Desc
+	DpiStatsApp       *prometheus.Desc
+	DpiStatsCat       *prometheus.Desc
+	DpiStatsRxBytes   *prometheus.Desc
+	DpiStatsRxPackets *prometheus.Desc
+	DpiStatsTxBytes   *prometheus.Desc
+	DpiStatsTxPackets *prometheus.Desc
+}
+
+func descClient(ns string) *uclient {
+	labels := []string{"name", "mac", "site_name", "gw_mac", "gw_name", "sw_mac", "sw_name", "vlan", "ip", "oui", "network", "sw_port", "wired"}
+	labelWireless := append([]string{"ap_mac", "ap_name", "radio_name", "radio", "radio_proto", "channel", "essid", "bssid", "radio_desc"}, labels...)
+	return &uclient{
+		Anomalies:      prometheus.NewDesc(ns+"anomalies_total", "Client Anomalies", labelWireless, nil),
```

Comment:
It looks like this isn't a counter.

```suggestion
		Anomalies:      prometheus.NewDesc(ns+"anomalies", "Client Anomalies", labelWireless, nil),
```

Reply:
it's a counter that resets. very odd. lets have a comment chain here. unless you're on discord?

I _just_ uploaded a new clients dashboard. Please grab it again. It'll be easier if you tell me what changes you make so i can make them on my dash. I'm editing a few things at once.....

Reply:
I also just realized some labels are screwed up in UAP :( gonna have to fix that up.

Reply:
Ok, I wonder what triggers the resets. It doesn't seem like a normal counter to me.

My changes:
* `[$__interval]` for rate, and setting a min step to 1m. This makes for nice zooming.
* Graph null as null, not connected
* Graph fill to 0
* Multiply byte rates by 8 to make them bps, since network people like to see Mbps not MBps.

Reply:
oh, also, you're probably going to want to run `1m` interval. newer unifi controller doesn't update every 30s...

Reply:
All of those options sound good except the part where we do math. I'd like to avoid that and just present the data as received from the controller. I'm trying to keep this as similar to the existing dashboard as possible. Let me see if I can find all those thingies you tweaked..

Reply:
This doesn't work? it just reduces no data points. `rate(unifipoller_client_transmit_bytes_total[$__interval])` but using `[5m]` does. This is the range vector it uses to find two points to calculate the rate. 5m seems reasonable and will overcome any missing points because the poller went down for a few minutes. no?

Reply:
You need to set a "min step" in order to make sure `$__interval` doesn't get smaller than your scrape interval.

![image](https://user-images.githubusercontent.com/1320667/69865095-9f9aa180-12a0-11ea-92a8-6717d1df7274.png)

It's too bad the Unifi data updates are so slow. I'm used to having much more data in Promehteus. :grin: I'm polling the APs ever 15s via SNMP right now.

Reply:
I'm not sure, that really doesn't sounds like the right thing to put into the range vector. Even 10 or 15m seems more reasonable than locking it to the artificial interval you're setting at 2m. 2m doesn't provide the best resolution: 1m does, and if you set the min step to 1m and try to use $__interval as the rate range vector it doesn't work because it can't find two points in 1m.

Reply:
Correct, if your scrape interval is 1m, 2m in the min step is the minimum step you can do. The `$__interval` is in Grafana for exactly the use case of auto-scaling the range vector based on the graph. The min step is the minimum, not a lock on the step.

If the graph is scaled out to a week, `$__interval` will be adjusted accordingly, and min step will no longer apply.